### PR TITLE
Allow temp storage to grow to fit the largest requested data

### PR DESCRIPTION
**Pull request by klashenriksson**
Some background for this change: For some AMP code I need to sample the same memory with two different dataloaders. One loader will always request less data than the other, and if that is sampled before the larger one, the larger data no longer fits in the temp storage. This PR changes so that we grow the temp storage if the data does not fit. Assuming the requested data sizes do not change over time, this will only reallocate at the start until the largest buffersize is sampled.

What do you think of this?

### CODE REVIEW

**Review by AliGhadirzadeh — CHANGES_REQUESTED** [review]
I'd suggest to create `_temp_storage` whenever there is a difference between `tot_size` and `_temp_storage.shape` (and not only when the size is smaller). 

Also, I think it would be nicer to name it "total_size" instead of "tot_size" to be more compatible with the rest of variable names :D

**Comment by klashenriksson** [comment]
> I'd suggest to create `_temp_storage` whenever there is a difference between `tot_size` and `_temp_storage.shape` (and not only when the size is smaller).
> 
> Also, I think it would be nicer to name it "total_size" instead of "tot_size" to be more compatible with the rest of variable names :D

I do not think recreating the buffer every time there is a size difference makes sense. For my use case that would cause reallocations every backprop. An alternative approach could be to store temp storage(s) in a dictionary and do a lookup for the particular size and create if it does not exist, but to me that also just creates a lookup that we do not need.

As I understand the code, the *current* role of the storage is to act as a reusable buffer we can use while gathering sample points, and that we later throw away when we actually store the samples in the output tensor (and afterwards we can not safely rely on the data given by the temporary storage). Given these constraints I think it is fastest to just allocate the biggest size and return slices.

**Comment by AliGhadirzadeh** [comment]
> > > I'd suggest to create `_temp_storage` whenever there is a difference between `tot_size` and `_temp_storage.shape` (and not only when the size is smaller).
> > > Also, I think it would be nicer to name it "total_size" instead of "tot_size" to be more compatible with the rest of variable names :D
> > 
> > 
> > I do not think recreating the buffer every time there is a size difference makes sense. For my use case that would cause reallocations every backprop. An alternative approach could be to store temp storage(s) in a dictionary and do a lookup for the particular size and create if it does not exist, but to me that also just creates a lookup that we do not need.
> > As I understand the code, the _current_ role of the storage is to act as a reusable buffer we can use while gathering sample points, and that we later throw away when we actually store the samples in the output tensor (and afterwards we can not safely rely on the data given by the temporary storage, i.e we do not own it afterwards). Given these constraints I think it is fastest to just allocate the biggest size and return slices.
> 
> Actually, this change is needed for my use case. I made a multi buffer memory loader that receives a list of tables and should sample from them all. The behavior should be the same as if the memories are concatenated into one buffer and samples are drawn from the new buffer. Therefore, I need to sample different sizes in every batch from each memory in the list. Making this change (the one I requested) will not affect the standard training since the total size (rollout length * rollout count) is not changed. It does not affect your use case since again you are changing the size in every bp_step. But it makes it possible for me to draw different batch sizes from different buffers. Alternatively, I can fix this at the loader side. I.e., I can always sample fixed size batches from all the memories and throw away samples not needed from each buffer. I am not sure if this makes the whole pipeline faster. On the other hand, depending on the "sampler" I may get unintended behaviors. As an example, I cannot cover the whole dataset in one training epoch even if a sampler is implemented to cover the whole dataset.

I may have misunderstood something here. I'll take a look at it later time tonight.

**Review by AliGhadirzadeh — APPROVED** [review]

**Comment by AliGhadirzadeh** [comment]
@klashenriksson Sorry that I did not read your code careful enough. It is also fine for my use case.

File: tests/test_table.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    sample_count = 100
+    counts = [256, 512]
+    seq_len = 3
+    for _ in range(sample_count):
+        for count in counts:
+            sample1 = table.sample(count, seq_len)
+            sample2 = table.sample(count, seq_len)
+
+            for key in sample1.keys():
+                col_samp_1: torch.Tensor = sample1[key]
```
[the remark this answers is not in the document]
Comment (tgolsson) [reply tests/test_table.py:61]: I think that's fine.

**Review by tgolsson — APPROVED** [review]
LGTM!

**Pull request merged**